components/file_system/src/lib.rs

```diff
@@ -335,6 +338,78 @@ pub fn sync_dir<P: AsRef<Path>>(path: P) -> io::Result<()> {
     File::open(path)?.sync_all()
 }
 
+fn walk_dir<F>(mut dir: fs::ReadDir, op: &mut F) -> io::Result<()>
+where
+    F: FnMut(&Path) -> io::Result<()>,
+{
+    dir.try_for_each(|f| {
+        if let Ok(f) = f && let Ok(meta) = f.metadata() {
+            if meta.is_dir() {
+                if let Ok(read) = f.path().read_dir() {
+                    walk_dir(read, op)?;
+                }
+            } else {
+                op(&f.path())?;
+            }
+        }
+        Ok(())
+    })
+}
+
+pub fn naive_dir_size<P: AsRef<Path>>(path: P) -> io::Result<usize> {
+    let mut size = 0;
+    walk_dir(path.as_ref().read_dir()?, &mut |f| {
+        if let Ok(meta) = f.metadata() {
+            size += meta.len() as usize;
+        }
+        Ok(())
+    })?;
+    Ok(size)
+}
+
+#[derive(Default)]
+pub struct DedupedDirSizeCalculator {
+    inodes: HashSet<u64>,
+}
+
+impl DedupedDirSizeCalculator {
+    pub fn size<P: AsRef<Path>>(&mut self, path: P) -> io::Result<usize> {
+        if let Ok(size) = self.size_imp(path.as_ref()) {
+            return Ok(size);
+        }
+        naive_dir_size(path)
+    }
+
+    pub fn clear(&mut self) {
+        self.inodes.clear();
+    }
+
+    fn size_imp(&mut self, path: &Path) -> io::Result<usize> {
+        #[cfg(unix)]
+        {
+            use std::os::unix::fs::MetadataExt;
+            let mut size = 0;
+            let dev = path.metadata()?.dev();
+            walk_dir(path.read_dir()?, &mut |f| {
+                if let Ok(meta) = f.metadata() {
```

Comment:
I'm afraid this may be too expansive as pd heartbeat is quite frequent.

Reply:
it's tested in demo version and seems fine. @BusyJay 
However, assuming we delete the obsolete tablet in-time, do we still need this? @tabokie

Reply:
The reason why hardlinks need to be excluded is there may be many files are essential the same. But if we can do cleanup in time, then the there should be not much same files, hence no need to solve hardlink.